### PR TITLE
perf(filter): bound token-count tokenizer batches

### LINKED ISSUE #1039: Bound TokenNumFilter tokenizer batches to reduce default-batch memory
Part of #1031.

## Motivation

`TokenNumFilter` is a batched operator with a default Data-Juicer batch size of 1,000 rows. It currently sends every uncached text in that batch to the Hugging Face tokenizer in one call and retains the complete nested token-ID output until all row counts have been written. Even when each text has an ordinary context length, the number of simultaneously live token IDs therefore grows with the full Data-Juicer batch.

This is different from a single oversized-document edge case: a normal 1,000-row batch with roughly 512-1,024 tokens per row is sufficient to retain hundreds of thousands to one million Python token IDs at once. Multiple workers amplify that per-worker peak.

## Proposed change

- Tokenize at most 128 uncached texts per internal call.
- Immediately reduce each encoded sequence to one scalar length.
- Release each `BatchEncoding` and its `input_ids` before starting the next tokenizer call.
- Stage only scalar counts and write stats after every internal batch succeeds.

The change preserves cached counts, uncached-row order, `add_special_tokens=False`, tokenizer exception type/message, stats update atomicity, sibling fields, return identity, min/max filtering, and all public configuration. It does not limit the Data-Juicer batch, individual text length, workers, or concurrency.

## Evidence

Baseline: `datajuicer/data-juicer` `main` at `b397cb295ee2ea876129f32ca4e77c0731b83986`.

### Controlled default-batch workload

The controlled tokenizer returns Hugging Face-shaped `input_ids` and `attention_mask` lists. Each row has 1,024 token IDs; row counts cover the 128-row boundary, a 512-row batch, and the default 1,000-row batch. Every RSS/allocation cell is the median of 5 fresh processes in randomized order. Count SHA-256 digests matched exactly at every size.

| Rows | Current eager RSS | Bounded RSS | RSS reduction | Current Python peak | Bounded Python peak | Python reduction |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 128 | 241.7 MiB | 241.3 MiB | 0.2% | 6.1 MiB | 6.1 MiB | 0.2% |
| 512 | 262.0 MiB | 242.9 MiB | 7.3% | 24.4 MiB | 6.2 MiB | 74.8% |
| 1,000 | 286.1 MiB | 243.1 MiB | 15.0% | 47.7 MiB | 6.2 MiB | 86.9% |

RSS growth fell from 53,270.88 to 2,084.37 bytes per row (-96.1%). Algorithm-local Python-allocation growth fell from 49,988.50 to 147.22 bytes per row (-99.7%). Maximum RSS sample CV was 2.62%.

### Real default Hugging Face tokenizer

The same A/B harness used `EleutherAI/pythia-6.9b-deduped`. Input rows produce 515-516 actual tokens each. Each memory cell is again the median of 5 fresh processes, and exact count digests matched.

| Rows | Current eager RSS | Bounded RSS | RSS reduction | Current Python peak | Bounded Python peak | Python reduction |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 128 | 553.5 MiB | 576.9 MiB | -4.2% | 3.2 MiB | 3.2 MiB | ~0% |
| 1,000 | 639.2 MiB | 562.0 MiB | 12.1% | 24.6 MiB | 3.3 MiB | 86.6% |

At 128 rows both implementations use one tokenizer call, so there is no mechanism-level allocation difference; the RSS reversal at that boundary is process/model baseline noise. At the default 1,000 rows, peak RSS decreased by 77.2 MiB. Maximum RSS sample CV was 5.49%, and real-tokenizer Python-allocation growth decreased by 99.4%.

### Latency

Real-tokenizer latency uses 256 rows with approximately 256 tokens per row, 3 fresh processes per variant, 40 observations per process, and 3 warmups (120 observations per variant). Process order was randomized.

| Current median | Bounded median | Throughput ratio | Bounded/current P95 |
| ---: | ---: | ---: | ---: |
| 34.232 ms | 31.072 ms | 1.102x | 1.081x |

Median throughput improved by 10.2%; P95 latency increased by 8.1%, within the normal 1.50x tail-latency budget.

Environment: Apple M2 Pro, 32 GiB RAM, macOS 26.2, Python 3.11.15. Fresh-process RSS includes interpreter and tokenizer/model state, so the matching algorithm-local allocation measurements identify the operator-owned mechanism directly.

Reproduction:

- [benchmark branch](https://github.com/macroguo-ghy/data-juicer/tree/benchmark/token-num-bounded-batch)
- [script pinned at the measured commit](https://github.com/macroguo-ghy/data-juicer/blob/9d79acf9095ec664d13e09f90ceea02a550b40d1/tests/benchmark_performance/benchmark_token_num_bounded_batch.py)
- Implementation PR: #1040

## Correctness coverage

- Exact batching boundaries at 0, 1, 127, 128, 129, 255, 256, and 257 uncached rows.
- Cached/uncached rows interleaved across internal batch boundaries without recomputation or reordering.
- Empty, ASCII, whitespace, Chinese, Unicode, and emoji text in 200 deterministic randomized differential cases.
- Auxiliary tokenizer outputs ignored exactly as before.
- A later internal tokenizer failure leaves all uncached stats untouched and preserves the error.
- Each encoded batch is released before the next tokenizer call.
- Fully cached batches bypass `get_model`; sibling columns, unrelated stats, and batch object identity are preserved.
- Existing tests use the real default tokenizer and verify exact counts and filtering decisions.

The focused suite contains 15 passing tests, including 13 cases added for this change.

## Remaining limits

- One arbitrarily long text can still require a complete exact token sequence from the external tokenizer.
- The tokenizer may still create unused auxiliary fields such as attention masks, but changing requested tokenizer outputs is a separate optimization and is intentionally not mixed into this PR.

### PR DESCRIPTION
**Pull request by macroguo-ghy**
## Summary

- Tokenize at most 128 uncached rows per internal call instead of retaining token IDs for the full Data-Juicer batch.
- Reduce each encoded batch to scalar counts and release it before the next tokenizer call; publish stats only after every call succeeds.
- Preserve cached counts, row order, tokenizer arguments, failure atomicity, sibling fields, and filtering behavior with boundary and randomized differential tests.

Closes #1039.
Part of #1031.

## Benchmark

The reproducible harness is intentionally maintained outside this PR so normal CI does not execute the memory campaign:

- [benchmark branch](https://github.com/macroguo-ghy/data-juicer/tree/benchmark/token-num-bounded-batch)
- [script pinned at the measured commit](https://github.com/macroguo-ghy/data-juicer/blob/9d79acf9095ec664d13e09f90ceea02a550b40d1/tests/benchmark_performance/benchmark_token_num_bounded_batch.py)

### Controlled default-batch workload

The controlled tokenizer returns Hugging Face-shaped `input_ids` and `attention_mask` lists, with 1,024 token IDs per row. Every memory cell is the median of 5 fresh processes in randomized order; exact count digests matched at every size.

| Rows | Current eager RSS | Bounded RSS | RSS reduction | Current Python peak | Bounded Python peak | Python reduction |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 128 | 241.7 MiB | 241.3 MiB | 0.2% | 6.1 MiB | 6.1 MiB | 0.2% |
| 512 | 262.0 MiB | 242.9 MiB | 7.3% | 24.4 MiB | 6.2 MiB | 74.8% |
| 1,000 | 286.1 MiB | 243.1 MiB | 15.0% | 47.7 MiB | 6.2 MiB | 86.9% |

RSS growth decreased by 96.1%, and algorithm-local Python-allocation growth decreased by 99.7%. Maximum RSS sample CV was 2.62%.

### Real default Hugging Face tokenizer

The A/B harness used `EleutherAI/pythia-6.9b-deduped`; input rows produced 515-516 actual tokens each. Every memory cell is the median of 5 fresh processes, exact count digests matched, and maximum RSS sample CV was 5.49%.

| Rows | Current eager RSS | Bounded RSS | RSS reduction | Current Python peak | Bounded Python peak | Python reduction |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 128 | 553.5 MiB | 576.9 MiB | -4.2% | 3.2 MiB | 3.2 MiB | ~0% |
| 1,000 | 639.2 MiB | 562.0 MiB | 12.1% | 24.6 MiB | 3.3 MiB | 86.6% |

At 128 rows both implementations make one tokenizer call, so there is no mechanism-level allocation difference; the RSS reversal is process/model baseline noise. At the default 1,000 rows, peak RSS decreased by 77.2 MiB, and Python-allocation growth decreased by 99.4%.

Real-tokenizer latency used 256 rows with approximately 256 tokens per row, 3 fresh processes per variant, 40 observations per process, and 3 warmups (120 observations per variant). Process order was randomized.

| Current median | Bounded median | Throughput ratio | Bounded/current P95 |
| ---: | ---: | ---: | ---: |
| 34.232 ms | 31.072 ms | 1.102x | 1.081x |

Median throughput improved by 10.2%; P95 latency increased by 8.1%, within the normal 1.50x tail-latency budget.

Environment: Apple M2 Pro, 32 GiB RAM, macOS 26.2, Python 3.11.15. Full methodology, applicability, and remaining limits are recorded in #1039.

## Testing

```bash
PYTHONPATH="$PWD" python -m pytest -q \
  tests/ops/filter/test_token_num_filter.py \
  tests/ops/filter/test_token_num_filter_bounded_batch.py

PYTHONPATH="$PWD" python -m black --check \
  data_juicer/ops/filter/token_num_filter.py \
  tests/ops/filter/test_token_num_filter_bounded_batch.py

PYTHONPATH="$PWD" python -m flake8 \
  --max-line-length=120 --extend-ignore=E203,E501,BLK100,F541 \
  data_juicer/ops/filter/token_num_filter.py \
  tests/ops/filter/test_token_num_filter_bounded_batch.py
```

- 15 targeted tests passed, including the existing real-default-tokenizer tests and 13 focused cases added by this PR.
- Boundaries cover 0, 1, 127, 128, 129, 255, 256, and 257 uncached rows.
- 200 deterministic randomized cases cover cached/uncached mixtures, empty and Unicode text, exact counts and order, auxiliary outputs, object identity, and unrelated fields.
- Failure tests prove that a later internal tokenizer error leaves all uncached stats untouched, and lifetime tests prove each encoded batch is released before the next call.

### CODE REVIEW

**Review by cmgzn — APPROVED** [review]
LGTM! Thanks for your work!
